### PR TITLE
Refactor StatusRepository

### PR DESCRIPTION
**Pull request by OscarSpruit**
Refactor `StatusRepository` to use flows and remove the complex polling mechanism.

### CODE REVIEW

File: drop-in/src/main/java/com/adyen/checkout/dropin/ActionHandler.kt [hunk 1]
```diff
@@ -83,10 +83,10 @@ class ActionHandler(
         activity: FragmentActivity,
         provider: ActionComponentProvider<out BaseActionComponent<out Configuration>, out Configuration>
     ) {
-        getActionComponentFor(activity, provider, dropInConfiguration).apply {
+        getActionComponentFor(activity, activity.application, provider, dropInConfiguration).apply {
```
Comment (etman55) [inline drop-in/src/main/java/com/adyen/checkout/dropin/ActionHandler.kt:86]: I guess this has nothing to do with StatusRepository refactor, so it's better to be done in a separate PR.

File: drop-in/src/main/java/com/adyen/checkout/dropin/ComponentParsingProvider.kt [hunk 2]
```diff
@@ -444,51 +445,45 @@ internal fun getActionProviderFor(
...
- * Provides a [BaseActionComponent] from an [ActionComponentProvider].
- *
- * @param activity The Action which the ActionComponent lifecycle will be bound to.
- * @param provider The component provider.
- * @throws CheckoutException In case a component cannot be created.
- */
 @Suppress("ComplexMethod", "LongMethod")
-internal fun getActionComponentFor(
-    activity: FragmentActivity,
+internal fun <T> getActionComponentFor(
```
Comment (etman55) [inline drop-in/src/main/java/com/adyen/checkout/dropin/ComponentParsingProvider.kt:449]: I guess this has nothing to do with StatusRepository refactor, so it's better to be done in a separate PR.

File: drop-in/src/main/java/com/adyen/checkout/dropin/ui/action/ActionComponentDialogFragment.kt [hunk 3]
```diff
@@ -147,13 +147,21 @@ class ActionComponentDialogFragment : DropInBottomSheetDialogFragment(), Observe
...
+
         if (!provider.requiresView(action)) {
             throw ComponentException(
                 "Action is not viewable - action: ${action.type} - " +
                     "paymentMethod: ${action.paymentMethodType}"
             )
         }
-        val component = getActionComponentFor(requireActivity(), provider, dropInViewModel.dropInConfiguration)
+
+        val component = getActionComponentFor(
```
Comment (etman55) [inline drop-in/src/main/java/com/adyen/checkout/dropin/ui/action/ActionComponentDialogFragment.kt:158]: I guess this has nothing to do with StatusRepository refactor, so it's better to be done in a separate PR.

File: drop-in/src/main/java/com/adyen/checkout/dropin/ActionHandler.kt (hunk 1 above)
Comment (OscarSpruit) [reply drop-in/src/main/java/com/adyen/checkout/dropin/ActionHandler.kt:86]: Fair enough, will keep that in mind for next time. It does however solve a big issue with the `StatusRepository`. The component wasn't destroyed after the fragment was, causing the polling to go on as long as the `DropInActivity` is alive

File: components-core/src/main/java/com/adyen/checkout/components/status/StatusRepository.kt [hunk 4]
```diff
@@ -1,142 +1,99 @@
...
-            elapsedTime <= MAX_POLLING_DURATION_MILLIS -> pollingDelay = POLLING_DELAY_SLOW
-            else -> _errorLiveData.setValue(ComponentException("Status requesting timed out with no result"))
-        }
+    fun refreshStatus(paymentData: String) {
+        refreshFlow.tryEmit(paymentData)
     }
 
     companion object {
-        val TAG = getTag()
         val MAX_POLLING_DURATION_MILLIS = TimeUnit.MINUTES.toMillis(15)
```
[the remark this answers is not in the document]
Comment (OscarSpruit) [reply components-core/src/main/java/com/adyen/checkout/components/status/StatusRepository.kt:94]: Also noticed that, but now that I know how the payment method works it makes sense. For QR codes for example, they have to finish the flow in another app. 15 minutes gives people plenty of time to complete the stuff in that app and come back to us

File: drop-in/src/main/java/com/adyen/checkout/dropin/ActionHandler.kt (hunk 1 above)
Comment (etman55) [reply drop-in/src/main/java/com/adyen/checkout/dropin/ActionHandler.kt:86]: good catch, I didn't notice that.

File: components-core/src/main/java/com/adyen/checkout/components/status/StatusRepository.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (etman55) [reply components-core/src/main/java/com/adyen/checkout/components/status/StatusRepository.kt:94]: Okay, but what happen when the user comes back to us does it start counting again or continue counting from last time?

[the remark this answers is not in the document]
Comment (OscarSpruit) [reply components-core/src/main/java/com/adyen/checkout/components/status/StatusRepository.kt:94]: It just continues and then when the time is over an error message is displayed

**Review by etman55 — APPROVED** [review]

**Review by jreij — APPROVED** [review]
LGTM 👍
As we discussed it would be nice if we can find a way to get the timer status from the repository itself but it's not so necessary
